JLine error when attempting to run the REPL with piped input

Scala 3.0.0-RC2. A command like the following:

```shell
$ scala <<EOF
:load somestuff.scala
EOF
```

or trying:

```shell
$ echo ":load src/script/scala/progscala3/hierarchy/Asserts.scala" | scala
```

Both cause this exception:

```
Exception in thread "main" java.lang.IllegalStateException: Unable to create a system terminal
	at org.jline.terminal.TerminalBuilder.doBuild(TerminalBuilder.java:323)
	at org.jline.terminal.TerminalBuilder.build(TerminalBuilder.java:265)
	at dotty.tools.repl.JLineTerminal.<init>(JLineTerminal.scala:23)
	at dotty.tools.repl.ReplDriver.runUntilQuit(ReplDriver.scala:123)
	at dotty.tools.repl.ReplDriver.tryRunning(ReplDriver.scala:114)
	at dotty.tools.repl.Main$.main(Main.scala:6)
	at dotty.tools.repl.Main.main(Main.scala)
```

This is fairly well known issue, see for example https://youtrack.jetbrains.com/issue/IDEA-183619. I attempted to use the recommended fix, setting the property `-Dorg.jline.terminal.dumb=true`. I did this by passing the argument `-J-Dorg.jline.terminal.dumb=true` to the REPL, even though this argument is not documented for the REPL in the `-help` output. It appeared to be ignored or do nothing. I also tried setting the environment variable `JAVA_OPTS=-Dorg.jline.terminal.dumb=true` before starting the REPL, but that didn't work either. Finally, I even tried setting the property inside the REPL, i.e.,

```shell
$ scala <<EOF
System.setProperty("...", "...")
:load somestuff.scala
EOF
```

I would expect this to be too late. It also doesn't help.

Note that in the Scala 2.13 REPL, without setting the property, you get a warning that it will default to a dumb terminal, but otherwise appears to work fine.

Sadly, this is still broken, even on the new `scala` runner / Scala CLI.

```bash
echo 'object Example { def smth = "smth" }' |scala-cli repl _   
# Compiling project (Scala 3.4.2, JVM (17))
# Compiled project (Scala 3.4.2, JVM (17))
# Exception in thread "main" java.lang.IllegalStateException: Unable to create a terminal
# 	at org.jline.terminal.TerminalBuilder.doBuild(TerminalBuilder.java:394)
# 	at org.jline.terminal.TerminalBuilder.build(TerminalBuilder.java:362)
# 	at dotty.tools.repl.JLineTerminal.<init>(JLineTerminal.scala:25)
# 	at dotty.tools.repl.ReplDriver.runUntilQuit(ReplDriver.scala:145)
# 	at dotty.tools.repl.ReplDriver.tryRunning(ReplDriver.scala:136)
# 	at dotty.tools.repl.Main$.main(Main.scala:7)
# 	at dotty.tools.repl.Main.main(Main.scala)
# 	Suppressed: java.io.IOException: Unable to find terminal provider ffm
# 		at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:74)
# 		at org.jline.terminal.TerminalBuilder.checkProvider(TerminalBuilder.java:667)
# 		at org.jline.terminal.TerminalBuilder.getProviders(TerminalBuilder.java:632)
# 		at org.jline.terminal.TerminalBuilder.doBuild(TerminalBuilder.java:395)
# 		... 6 more
# 	Suppressed: java.io.IOException: Unable to find terminal provider jni
# 		at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:74)
# 		at org.jline.terminal.TerminalBuilder.checkProvider(TerminalBuilder.java:667)
# 		at org.jline.terminal.TerminalBuilder.getProviders(TerminalBuilder.java:634)
# 		at org.jline.terminal.TerminalBuilder.doBuild(TerminalBuilder.java:395)
# 		... 6 more
# 	Suppressed: java.io.IOException: Unable to find terminal provider jansi
# 		at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:74)
# 		at org.jline.terminal.TerminalBuilder.checkProvider(TerminalBuilder.java:667)
# 		at org.jline.terminal.TerminalBuilder.getProviders(TerminalBuilder.java:636)
# 		at org.jline.terminal.TerminalBuilder.doBuild(TerminalBuilder.java:395)
# 		... 6 more
# 	Suppressed: java.io.IOException: Unable to load terminal provider jna: null
# 		at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:71)
# 		at org.jline.terminal.TerminalBuilder.checkProvider(TerminalBuilder.java:667)
# 		at org.jline.terminal.TerminalBuilder.getProviders(TerminalBuilder.java:638)
# 		at org.jline.terminal.TerminalBuilder.doBuild(TerminalBuilder.java:395)
# 		... 6 more
# 	Caused by: java.lang.reflect.InvocationTargetException
# 		at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
# 		at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
# 		at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
# 		at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
# 		at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:480)
# 		at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:69)
# 		... 9 more
# 	Caused by: java.lang.UnsupportedOperationException: Unsupported platform mac-aarch64
# 		at org.jline.terminal.impl.jna.JnaNativePty.isatty(JnaNativePty.java:194)
# 		at org.jline.terminal.impl.jna.JnaNativePty.isPosixSystemStream(JnaNativePty.java:170)
# 		at org.jline.terminal.impl.jna.JnaTerminalProvider.checkSystemStream(JnaTerminalProvider.java:124)
# 		at org.jline.terminal.impl.jna.JnaTerminalProvider.<init>(JnaTerminalProvider.java:31)
# 		... 15 more
```

--- a/compiler/src/dotty/tools/repl/JLineTerminal.scala
+++ b/compiler/src/dotty/tools/repl/JLineTerminal.scala
@@ -21,11 +21,16 @@ class JLineTerminal extends java.io.Closeable {
   // Logger.getLogger("org.jline").setLevel(Level.FINEST)
 
   private val terminal =
-    TerminalBuilder.builder()
-    .dumb(dumbTerminal) // fail early if not able to create a terminal
-    .build()
+    var builder = TerminalBuilder.builder()
+    if System.getenv("TERM") == "dumb" then
+      // Force dumb terminal if `TERM` is `"dumb"`.
+      // Note: the default value for the `dumb` option is `null`, which allows
+      // JLine to fall back to a dumb terminal. This is different than `true` or
+      // `false` and can't be set using the `dumb` setter.
+      // This option is used at https://github.com/jline/jline3/blob/894b5e72cde28a551079402add4caea7f5527806/terminal/src/main/java/org/jline/terminal/TerminalBuilder.java#L528.
+      builder.dumb(true)
+    builder.build()
   private val history = new DefaultHistory
-  def dumbTerminal = Option(System.getenv("TERM")) == Some("dumb")
 
   private def blue(str: String)(using Context) =
     if (ctx.settings.color.value != "never") Console.BLUE + str + Console.RESET

--- a/compiler/test-coursier/dotty/tools/coursier/CoursierScalaTests.scala
+++ b/compiler/test-coursier/dotty/tools/coursier/CoursierScalaTests.scala
@@ -75,8 +75,8 @@ class CoursierScalaTests:
     version()
 
     def emptyArgsEqualsRepl() =
-      val output = CoursierScalaTests.csScalaCmd()
-      assertTrue(output.mkString("\n").contains("Unable to create a terminal")) // Scala attempted to create REPL so we can assume it is working
+      val output = CoursierScalaTests.csScalaCmdWithStdin(Seq.empty, Some("println(\"Hello World\")\n:quit"))
+      assertTrue(output.mkString("\n").contains("Hello World"))
     emptyArgsEqualsRepl()
 
     def run() =
@@ -132,8 +132,8 @@ class CoursierScalaTests:
     compileFilesToJarAndRun()
 
     def replWithArgs() =
-      val output = CoursierScalaTests.csScalaCmd("-source", "3.0-migration")
-      assertTrue(output.mkString("\n").contains("Unable to create a terminal")) // Scala attempted to create REPL so we can assume it is working
+      val output = CoursierScalaTests.csScalaCmdWithStdin(Seq("-source", "3.0-migration"), Some("println(\"Hello World\")\n:quit"))
+      assertTrue(output.mkString("\n").contains("Hello World"))
     replWithArgs()
 
     def argumentFile() =
@@ -148,25 +148,31 @@ class CoursierScalaTests:
 
 object CoursierScalaTests:
 
-  def execCmd(command: String, options: String*): (Int, List[String]) =
+  private def execCmd(command: String, options: Seq[String] = Seq.empty, stdin: Option[String] = None): (Int, List[String]) =
     val cmd = (command :: options.toList).toSeq.mkString(" ")
     val out = new ListBuffer[String]
-    val code = cmd.!(ProcessLogger(out += _, out += _))
+    val process = stdin match
+      case Some(input) => Process(cmd) #< new java.io.ByteArrayInputStream(input.getBytes)
+      case None => Process(cmd)
+    val code = process.!(ProcessLogger(out += _, out += _))
     (code, out.toList)
 
   def csScalaCmd(options: String*): List[String] =
-    csCmd("dotty.tools.MainGenericRunner", options*)
+    csScalaCmdWithStdin(options, None)
+  
+  def csScalaCmdWithStdin(options: Seq[String], stdin: Option[String]): List[String] =
+    csCmd("dotty.tools.MainGenericRunner", options, stdin)
 
   def csScalaCompilerCmd(options: String*): List[String] =
-    csCmd("dotty.tools.dotc.Main", options*)
+    csCmd("dotty.tools.dotc.Main", options)
 
-  private def csCmd(entry: String, options: String*): List[String] =
+  private def csCmd(entry: String, options: Seq[String], stdin: Option[String] = None): List[String] =
     val (jOpts, args) = options.partition(_.startsWith("-J"))
     val newOptions = args match
       case Nil => args
       case _ => "--" +: args
     val newJOpts = jOpts.map(s => s"--java-opt ${s.stripPrefix("-J")}").mkString(" ")
-    execCmd("./cs", (s"""launch "org.scala-lang:scala3-compiler_3:${sys.env("DOTTY_BOOTSTRAPPED_VERSION")}" $newJOpts --main-class "$entry" --property "scala.usejavacp=true" --property "scala.use_legacy_launcher=true"""" +: newOptions)*)._2
+    execCmd("./cs", (s"""launch "org.scala-lang:scala3-compiler_3:${sys.env("DOTTY_BOOTSTRAPPED_VERSION")}" $newJOpts --main-class "$entry" --property "scala.usejavacp=true" --property "scala.use_legacy_launcher=true"""" +: newOptions), stdin)._2
 
   /** Get coursier script */
   @BeforeClass def setup(): Unit =
@@ -177,7 +183,7 @@ object CoursierScalaTests:
       case other => fail(s"Unsupported OS for coursier launcher: $other")
 
     def runAndCheckCmd(cmd: String, options: String*): Unit =
-      val (code, out) = execCmd(cmd, options*)
+      val (code, out) = execCmd(cmd, options)
       if code != 0 then
         fail(s"Failed to run $cmd ${options.mkString(" ")}, exit code: $code, output: ${out.mkString("\n")}")
 

